ydb/library/yql/minikql/comp_nodes/mkql_wide_combine.cpp

```diff
@@ -302,6 +315,425 @@ class TState : public TComputationValue<TState> {
     TStates States;
 };
 
+class TSpillingSupportState : public TComputationValue<TSpillingSupportState> {
+    typedef TComputationValue<TSpillingSupportState> TBase;
+    typedef std::optional<NThreading::TFuture<ISpiller::TKey>> TAsyncWriteOperation;
+    typedef std::optional<NThreading::TFuture<std::optional<TRope>>> TAsyncReadOperation;
+
+    struct TSpilledBucket {
+        std::unique_ptr<TWideUnboxedValuesSpillerAdapter> SpilledState; //state collected before switching to spilling mode
+        std::unique_ptr<TWideUnboxedValuesSpillerAdapter> SpilledData; //data collected in spilling mode
+        std::unique_ptr<TState> InMemoryProcessingState;
+        TAsyncWriteOperation AsyncWriteOperation;
+
+        enum class EBucketState {
+            InMemory,
+            SpillingState,
+            SpillingData
+        };
+
+        EBucketState BucketState = EBucketState::InMemory;
+    };
+public:
+    enum class EOperatingMode {
+        InMemory,
+        Spilling,
+        ProcessSpilled
+    };
+    TSpillingSupportState(
+        TMemoryUsageInfo* memInfo,
+        const TCombinerNodes& nodes, IComputationWideFlowNode *const flow, size_t wideFieldsIndex,
+        const TMultiType* usedInputItemType, const TMultiType* keyAndStateType, ui32 keyWidth,
+        const THashFunc& hash, const TEqualsFunc& equal
+    )
+        : TBase(memInfo)
+        , InMemoryProcessingState(memInfo, keyWidth, keyAndStateType->GetElementsCount() - keyWidth, hash, equal)
+        , Nodes(nodes)
+        , Flow(flow)
+        , WideFieldsIndex(wideFieldsIndex)
+        , UsedInputItemType(usedInputItemType)
+        , KeyAndStateType(keyAndStateType)
+        , KeyWidth(keyWidth)
+        , Hasher(hash)
+        , Mode(EOperatingMode::InMemory)
+        , MemInfo(memInfo)
+        , Equal(equal)
+    {
+        BufferForUsedInputItems.reserve(usedInputItemType->GetElementsCount());
+        BufferForKeyAnsState.reserve(keyAndStateType->GetElementsCount());
+    }
+    ~TSpillingSupportState() {
+    }
+
+    EFetchResult DoCalculate(TComputationContext& ctx, NUdf::TUnboxedValue*const* output) {
+        while (true) {
+            switch(GetMode()) {
+                case EOperatingMode::InMemory: {
+                    auto r = DoCalculateInMemory(ctx, output);
+                    if (GetMode() == TSpillingSupportState::EOperatingMode::InMemory) {
```

Comment:
Why cannot simplify this to smth like
```
return DoCalculateInMemory(ctx, output);;
```

Reply:
During the execution of DoCalculateInMemory state may be switched to Spilling. In case if the state didn't change we want to return the result. In case if state is Spilling we want to continue processing but in different state